realm/realm-library/src/androidTest/java/io/realm/DynamicRealmAsyncQueryTests.kt

```diff
@@ -0,0 +1,370 @@
+    val configFactory = TestRealmConfigurationFactory()
+    @get:Rule
+    val thrown: ExpectedException = ExpectedException.none()
+
+    private val looperThread = BlockingLooperThread()
+
+    private lateinit var config: RealmConfiguration
+
+    @Before
+    fun setUp() {
+        config = configFactory.createConfiguration()
+
+        // Initializes schema. DynamicRealm will not do that, so let a normal Realm create the file first.
+        Realm.getInstance(config).close()
+    }
+
+    // ****************************
+    // ****  Async transaction  ***
+    // ****************************
+    // Starts asynchronously a transaction to insert one element.
+    @Test
+    fun executeTransactionAsync() = looperThread.runBlocking {
+        val realm = DynamicRealm.getInstance(config)
+                .also { looperThread.closeAfterTest(it) }
+
+        assertEquals(0, realm.where(Owner.CLASS_NAME).count())
+
+        realm.executeTransactionAsync({ transactionRealm ->
+            val owner = transactionRealm.createObject(Owner.CLASS_NAME)
+            owner.setString(Owner.FIELD_NAME, "Owner")
+        }, {
+            assertEquals(1, realm.where(Owner.CLASS_NAME).count())
+            assertEquals("Owner", realm.where(Owner.CLASS_NAME).findFirst()!!.getString(Owner.FIELD_NAME))
+            looperThread.testComplete()
+        }) { error ->
+            fail(error.message)
+        }
+    }
+
+    @Test
+    fun executeTransactionAsync_onSuccess() = looperThread.runBlocking {
+        val realm = DynamicRealm.getInstance(config)
+                .also { looperThread.closeAfterTest(it) }
+
+        assertEquals(0, realm.where(Owner.CLASS_NAME).count())
+
+        realm.executeTransactionAsync(DynamicRealm.Transaction { transactionRealm ->
+            val owner = transactionRealm.createObject(Owner.CLASS_NAME)
+            owner.setString(Owner.FIELD_NAME, "Owner")
+        }, DynamicRealm.Transaction.OnSuccess {
+            assertEquals(1, realm.where(Owner.CLASS_NAME).count())
+            assertEquals("Owner", realm.where(Owner.CLASS_NAME).findFirst()!!.getString(Owner.FIELD_NAME))
+            looperThread.testComplete()
+        })
+    }
+
+    @Test
+    fun executeTransactionAsync_onSuccessCallerRealmClosed() = looperThread.runBlocking {
+        val realm = DynamicRealm.getInstance(config)
+
+        assertEquals(0, realm.where(Owner.CLASS_NAME).count())
+        realm.executeTransactionAsync(DynamicRealm.Transaction { transactionRealm ->
+            val owner = transactionRealm.createObject(Owner.CLASS_NAME)
+            owner.setString(Owner.FIELD_NAME, "Owner")
+        }, DynamicRealm.Transaction.OnSuccess {
+            assertTrue(realm.isClosed)
+            val newRealm = DynamicRealm.getInstance(config)
+            assertEquals(1, newRealm.where(Owner.CLASS_NAME).count())
+            assertEquals("Owner", newRealm.where(Owner.CLASS_NAME).findFirst()!!.getString(Owner.FIELD_NAME))
+
+            newRealm.close()
```

Comment:
I would do `DynamicRealm.getInstance(config).use { newRealm -> ... }` . Otherwise, if the assserts fail, this code is not called.